### ISSUE TITLE
How do you get the current path?

### ISSUE THREAD

**Issue by snarlynarwhal** [inkle/ink-unity-integration#38]
I tried most properties that have 'string' or 'path' in the name, but most return empty strings or null references. It seems that story.path.ToString() would would work, but it doesn't. I just want to know what knot/stitch the story is currently at.

**Comment by joethephish** [comment]
This is quite a common question, but the answer is slightly less useful I'm afraid. If you're at the end of a section (i.e. at a choice point) then strictly speaking, the story doesn't actually *have* a path, since it's "fallen off the end of the story". Take this story for example:

```
Hello world
<- some_choices
-> DONE

== some_choices
* hi -> elsewhere
* there -> elsewhere
```

Once you've run the story to the first set of choices, it's inserted them from `some_choices`, returned back to the main story, hit the `-> DONE`. Where is the story exactly? Well, nowhere.

Depending on what you're actually interested in, you could take one of the `Choice` objects that you have, and get the `sourcePath` property on it - in the above case that would give you a path with `some_choices` in it.

Potentially we could allow you to get the current path when the story's `canContinue` property returns true, since it's still in the middle of "line by line evaluation mode". But I'd suggest that perhaps you need to solve your problem a different way?

**Comment by snarlynarwhal** [comment]
I realized I wanted a very peculiar game flow that I realized using C# and Ink tags. In case you're curious:
1. Store numerous storyline paths in a list and shuffle the list.
2. Choose a storyline from the list at random and remove it from the list.
3. Divert the story to that path and continue the storyline until reaching a path tag
4. Store the path tag in the storyline list for continuing this storyline later on.
5. Repeat steps 2-4 a set number of times, then play an end turn storyline.
6. Repeat until the player reaches an end game scenario.

**Comment by joningold** [comment]
FWIW, I would do all this inside ink. Ink can stores paths and divert
targets in a more human readable way.

Jon

**Comment by joethephish** [comment]
Would be interested to see your ink solution for this @joningold!

**Comment by joningold** [comment]
Challenge accepted! The following associates content knots with list
entries, and so you do your logic on list entries, and just go to the
relevant knot when necessary.

I didn't quite understand the design with "hitting path tags and saving for
later", but note that if you wanted to leave a scene halfway and pick it up
again, you just make a new list entry (HalfEatenPorridge) which diverts
into the middle of the eating_porridge knot, and *don't add it to the
StoryBeats list until you've eaten half a bowl of porridge.*

tl;dr - each list item is a card, on each card is content knot, you add
cards when you need to, remove them when you pick them.

jon

===

LIST StoryBeats = (MeetTheBears), (EatThePorridge), (RunLikeCrazy)

-> gameLoop

=== gameLoop
~ temp finishedGame = false
- (top)
    { finishedGame:
        -> endContent
    - else:
        -> startNewStoryBeat -> top
    }

=== startNewStoryBeat
    ~ temp contentKnot = getNewStoryBeat()
    -> contentKnot ->->


=== function getNewStoryBeat()
    ~ temp beat = LIST_RANDOM(StoryBeats)
    ~ StoryBeats -= beat
    { not beat:
        >>> ERROR! Out of story content. Callin' that a win.
        ~ return -> endContent
    }
    ~ return knotForStoryBeat(beat)

=== function knotForStoryBeat(beat)
    { beat:
    - MeetTheBears :
        ~ return -> meet_the_bears
    - EatThePorridge:
        ~ return -> eat_porridge
    - RunLikeCrazy:
        ~ return -> run_like_crazy
    - else:
        >>> ERROR: No story content for beat {beat}. Calling it a win.
    }

// content knots

=== run_like_crazy
    *   Runnin'
        ->->
=== eat_porridge
    *   Eatin'
        ->->
=== meet_the_bears
    *   Meetin'
        ->->
=== endContent
    *   Endin'
        -> END

=== function LIST_RANDOM(list)
    { list:
        ~ temp k = RANDOM(1, LIST_COUNT(list))
        ~ return LIST_ENTRY_INDEX(list, k)
    - else:
        ~ return ()
    }

=== function LIST_ENTRY_INDEX(list, k)
    { k == 1:
        ~ return LIST_MIN(list)
    - else:
        ~ k--
        ~ list = list - LIST_MIN(list)
        ~ return LIST_ENTRY_INDEX(list, k)
    }

**Comment by joethephish** [comment]
Yeah, was fiddling myself and coming up with something similar. Then I hit the snag that we haven't actually defined a `LIST_RANDOM` (though I see you've nabbed that recursive one from Heaven's Vault... we should probably implement it properly since it seems generally useful, and I imagine that version is a bit slow).

**Comment by joethephish** [comment]
Oh and I think what @snarlynarwhal meant by "hitting path tags and saving for later" would be implemented something like:

```
// Only unlock MeetTheBears by default
LIST StoryBeats = (MeetTheBears), EatThePorridge, RunLikeCrazy

// Then, unlock stories at the end of other stories:
=== meet_the_bears
    *   Meetin'
        ~ StoryBeats += (EatThePorridge, RunLikeCrazy)
        ->->
```

**Comment by snarlynarwhal** [comment]
Oh wow thanks for the speedy replies and help guys, I appreciate it very much! @joethephish, that's exactly what I meant and @joningold, you nailed it! Thanks for taking the time to write such an extensive example for how I can solve my problem within Ink - definitely prefer to keep the story logic inside Ink.

PS - I just started using Ink last weekend and have been loving it so far!
